Sources/_SwiftSyntaxCShims/wasm_support.c

```diff
@@ -0,0 +1,35 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(__wasm32__)
+
+// uint32_t
+#define SWIFT_WASM_MACRO_ABI 1
+
+#define _STR(X) #X
+#define STR(X) _STR(X)
+
+// LLVM has special-cased handling to map .custom_section.foo to
+// Wasm Custom Section "foo". this must be a metadata section rather
+// than a data section so we can't use __attribute__((section)) for it.
+// See: https://reviews.llvm.org/D43097
+__asm__("\t.section .custom_section.swift_wasm_macro_abi,\"\",@\n\t.4byte " STR(SWIFT_WASM_MACRO_ABI) "\n");
```

Comment:
This adds a custom `swift_wasm_macro_abi` section to the binary with a little-endian uint32 value. We check the ABI version in the wasm executor in order to future-proof. See:

https://github.com/apple/swift/pull/73031/files#diff-2261ec558c289a4bd568f211772e906caa997f8c26b38e4c6d5d37b7b1554379R47

Will add a brief explanation to the comment.

Reply:
hmm actually @kateinoigakukun what do you think about using the export name to indicate the ABI instead? E.g. we could export the pump function with the name `swift_wasm_macro_pump_v1` and the runtime could check for the presence of this to indicate the v1 ABI. The custom section support feels shaky to me — I wouldn't be surprised if there are tools that don't know how to handle custom wasm sections (or, alternatively, strip them out.)

Reply:
Pushed this as a tentative change, happy to revert if the Custom Sections approach is preferable but I do like this atm because it's a bit more lightweight.

Reply:
For posterity: discussed this with @kateinoigakukun [here](https://swift-open-source.slack.com/archives/C06MUFU7ZCZ/p1715394639346069?thread_ts=1711575681.096789&cid=C06MUFU7ZCZ); we aligned on using the exported function name rather than a custom section, since some Wasm tooling (_ahem_ wasm-strip) can be too trigger-happy about stripping custom sections.

Reply:
I think Swift tries to be conservative with custom sections because it's a terribly non-portable solution. Although portability isn't at issue here, it might maybe make sense to use an exported function name. That'd be the typical solution here if we were talking about a C function on another platform, I think.